jupyterhub/apihandlers/base.py

```diff
@@ -201,19 +260,37 @@ def user_model(self, user, include_servers=False, include_state=False):
             'created': isoformat(user.created),
             'last_activity': isoformat(user.last_activity),
         }
-        if '' in user.spawners:
-            model['pending'] = user.spawners[''].pending
-
-        if not include_servers:
-            model['servers'] = None
-            return model
-
-        servers = model['servers'] = {}
-        for name, spawner in user.spawners.items():
-            # include 'active' servers, not just ready
-            # (this includes pending events)
-            if spawner.active:
-                servers[name] = self.server_model(spawner, include_state=include_state)
+        access_map = {
+            'read:users': set(model.keys()),  # All available components
+            'read:users:names': {'kind', 'name'},
+            'read:users:groups': {'kind', 'name', 'groups'},
+            'read:users:activity': {'kind', 'name', 'last_activity'},
+            'read:users:servers': {'kind', 'name', 'servers'},
+        }
+        self.log.debug(
+            "Asking for user models with scopes [%s]" % ", ".join(self.raw_scopes)
+        )
+        allowed_keys = set()
+        for scope in access_map:
+            if scope in self.parsed_scopes:
+                scope_filter = self.get_scope_filter(scope)
+                if scope_filter(user.name):
+                    allowed_keys |= access_map[scope]
+        model = {key: model[key] for key in allowed_keys if key in model}
+        if model:
+            if '' in user.spawners and 'pending' in allowed_keys:
+                model['pending'] = user.spawners[''].pending
+            if not (include_servers and 'servers' in allowed_keys):
+                model['servers'] = None
+            else:
+                servers = model['servers'] = {}
+                for name, spawner in user.spawners.items():
+                    # include 'active' servers, not just ready
+                    # (this includes pending events)
+                    if spawner.active:
+                        servers[name] = self.server_model(
+                            spawner, include_state=include_state
```

Comment:
For future, this `include_state` is the argument that I think should be resolved by a scope like the rest. Such as `read:user:auth_state` or perhaps `admin:user:auth_state` since it's a highly sensitive field that users shouldn't generally have access to themselves by default. That doesn't need to be in this PR, though. This is one of the things where we are using `user.admin` as a blunt proxy for what should be handled with scopes.